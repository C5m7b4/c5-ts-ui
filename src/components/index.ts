import {
  Button,
  NormalButtonType,
  NormalButtonSize,
  ButtonProps,
} from './Button';

export { Button, NormalButtonType, NormalButtonSize, ButtonProps };
