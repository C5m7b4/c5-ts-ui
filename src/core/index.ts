import Backdrop, { BackdropProps } from './Backdrop';
import Modal, { ModalProps } from './Modal';
import ModalWrapper, { ModalWrapperProps } from './ModalWrapper';

export {
  Backdrop,
  BackdropProps,
  Modal,
  ModalProps,
  ModalWrapper,
  ModalWrapperProps,
};
