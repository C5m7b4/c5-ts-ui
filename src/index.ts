import { Add } from './math';

export { Add };
