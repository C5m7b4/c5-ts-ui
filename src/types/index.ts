export type Theme = 'light' | 'dark' | 'colored';

export type Parsers =
  | 'babel'
  | 'css'
  | 'json'
  | 'flow'
  | 'babel-flow'
  | 'babel-ts'
  | 'typescript'
  | 'json'
  | 'markdown'
  | 'html';
