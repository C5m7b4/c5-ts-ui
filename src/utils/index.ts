export * from './colors';
export * from './reserved';
export * from './fns';
