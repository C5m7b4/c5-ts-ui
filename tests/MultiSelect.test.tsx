import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { MultiSelect } from '../src';
import { data } from './data';
import { backspace } from './helper';
import { act } from 'react-dom/test-utils';

export const createBubbledEvent = (type, props = {}) => {
  const event = new Event(type, { bubbles: true });
  Object.assign(event, props);
  return event;
};

describe('MultiSelect tests', () => {
  const setCheckedItems = jest.fn();

  test('should render correctcly', () => {
    const { container } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        id="id"
        selectBoxId="selectBoxId"
        backgroundColor="red"
        selectId="selectId"
        inputId="inputId"
      />
    );

    expect(container).toMatchSnapshot();
  });

  test('should be callable without an id', () => {
    const { container } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        selectBoxId="selectBoxId"
        backgroundColor="red"
        selectId="selectId"
        inputId="inputId"
      />
    );

    expect(container).toMatchSnapshot();
  });

  test('should be callable withough a selectBoxId', () => {
    const { container } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        backgroundColor="red"
        selectId="selectId"
        inputId="inputId"
      />
    );

    expect(container).toMatchSnapshot();
  });

  test('should be callable without a backgroundColor, selectId, or inputId', () => {
    const { container } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test('should be clickable', () => {
    jest.useFakeTimers();
    const { container } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
      />
    );

    const option = screen.getByRole('option', { name: /select a department/i });
    fireEvent.click(option);
    jest.advanceTimersByTime(1000);
    const checkBoxes = document.querySelector(
      '.multi-select-checkboxes'
    ) as HTMLElement;
    expect(checkBoxes).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('should be able to click on the first checkbox', async () => {
    jest.useFakeTimers();
    const { container } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
      />
    );

    const option = screen.getByRole('option', { name: /select a department/i });
    fireEvent.click(option);
    jest.advanceTimersByTime(1000);

    const check = screen.getByRole('checkbox', { name: /Grocery/i });
    expect(check).not.toBeChecked();

    await fireEvent.click(check);
    expect(check).toBeChecked();

    expect(container).toMatchSnapshot();
  });

  test('should close checkboxes window', () => {
    jest.useFakeTimers();
    render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
      />
    );

    const option = screen.getByRole('option', { name: /select a department/i });
    fireEvent.click(option);
    jest.advanceTimersByTime(1000);

    // now close the box
    fireEvent.click(option);
    const checkBoxes = document.querySelector(
      '.multi-select-checkboxes'
    ) as HTMLElement;
    expect(checkBoxes).not.toBeInTheDocument();
  });

  test('should handle unchecking a box', async () => {
    jest.useFakeTimers();
    render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
      />
    );

    const option = screen.getByRole('option', { name: /select a department/i });
    fireEvent.click(option);
    jest.advanceTimersByTime(1000);

    const check = screen.getByRole('checkbox', { name: /Grocery/i });
    expect(check).not.toBeChecked();

    await fireEvent.click(check);
    expect(check).toBeChecked();

    await fireEvent.click(check);
    expect(check).not.toBeChecked();

    // this test is to trigger line 109
    const check1 = screen.getByRole('checkbox', { name: /Nonfood/i });
    expect(check1).not.toBeChecked();

    await fireEvent.click(check1);
    expect(check1).toBeChecked();
  });

  test('should be able to search', () => {
    jest.useFakeTimers();
    render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        inputId={'inputId'}
      />
    );

    const option = screen.getByRole('option', { name: /select a department/ });
    fireEvent.click(option);
    jest.advanceTimersByTime(1000);

    const input = screen.getByPlaceholderText(/search/i) as HTMLInputElement;

    fireEvent.change(input, { target: { value: 'gro' } });
    expect(input.value).toEqual('gro');

    backspace(input);
    expect(input.value).toEqual('gr');

    backspace(input);
    backspace(input);
    expect(input.value).toEqual('');
  });

  test('should handle clicking when clientX < rect.left', () => {
    render(
      <div id="main" data-testid="main" style={{ padding: '100px' }}>
        <p>Here is the first div</p>
        <button>button</button>
        <MultiSelect
          data={data}
          defaultOption="select a department"
          keyDescriptor="keyDescriptor"
          optionValue="id"
          optionText="value"
          checkedItems={[]}
          setCheckedItems={setCheckedItems}
          inputId={'inputId'}
        />
      </div>
    );

    const option = screen.getByRole('option', { name: /select a department/ });
    fireEvent.click(option);
    jest.advanceTimersByTime(1000);

    const div = screen.queryByText(/button/i) as HTMLElement;

    act(() => {
      div.dispatchEvent(
        createBubbledEvent('click', { clientX: -5, clientY: 0 })
      );
    });
  });

  test('should handle click when clientX > rect.right', () => {
    render(
      <div id="main" data-testid="main" style={{ padding: '100px' }}>
        <p>Here is the first div</p>
        <button>button</button>
        <MultiSelect
          data={data}
          defaultOption="select a department"
          keyDescriptor="keyDescriptor"
          optionValue="id"
          optionText="value"
          checkedItems={[]}
          setCheckedItems={setCheckedItems}
          inputId={'inputId'}
        />
      </div>
    );

    const option = screen.getByRole('option', { name: /select a department/ });
    fireEvent.click(option);
    jest.advanceTimersByTime(1000);

    const div = screen.queryByText(/button/i) as HTMLElement;

    act(() => {
      div.dispatchEvent(
        createBubbledEvent('click', { clientX: 10, clientY: 0 })
      );
    });
  });

  test('should handle click when clientY < rect.top', () => {
    render(
      <div id="main" data-testid="main" style={{ padding: '100px' }}>
        <p>Here is the first div</p>
        <button>button</button>
        <MultiSelect
          data={data}
          defaultOption="select a department"
          keyDescriptor="keyDescriptor"
          optionValue="id"
          optionText="value"
          checkedItems={[]}
          setCheckedItems={setCheckedItems}
          inputId={'inputId'}
        />
      </div>
    );

    const option = screen.getByRole('option', { name: /select a department/ });
    fireEvent.click(option);
    jest.advanceTimersByTime(1000);

    const div = screen.queryByText(/button/i) as HTMLElement;

    act(() => {
      div.dispatchEvent(
        createBubbledEvent('click', { clientX: 0, clientY: -5 })
      );
    });
  });

  test('should handle click when clientY > rect.bottom', () => {
    render(
      <div id="main" data-testid="main" style={{ padding: '100px' }}>
        <p>Here is the first div</p>
        <button>button</button>
        <MultiSelect
          data={data}
          defaultOption="select a department"
          keyDescriptor="keyDescriptor"
          optionValue="id"
          optionText="value"
          checkedItems={[]}
          setCheckedItems={setCheckedItems}
          inputId={'inputId'}
        />
      </div>
    );

    const option = screen.getByRole('option', { name: /select a department/ });
    fireEvent.click(option);
    jest.advanceTimersByTime(1000);

    const div = screen.queryByText(/button/i) as HTMLElement;

    act(() => {
      div.dispatchEvent(
        createBubbledEvent('click', { clientX: 0, clientY: 20 })
      );
    });
  });
});
