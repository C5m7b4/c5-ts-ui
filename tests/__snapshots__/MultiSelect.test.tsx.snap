// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect tests should be able to click on the first checkbox 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="select-box-id"
    >
      <select
        class="form-control"
        id="select-id"
      >
        <option
          style="display: none;"
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    >
      <div
        class="multi-select-checkboxes"
        id="multi-select-checkboxes"
        style="top: 0px; left: 0px; background-color: rgb(31, 30, 30);"
      >
        <input
          class="form-control"
          id="multi-select-search"
          placeholder="Search"
          type="text"
          value=""
        />
        <label>
          <input
            type="checkbox"
          />
          Grocery
        </label>
        <label>
          <input
            type="checkbox"
          />
          Nonfood
        </label>
        <label>
          <input
            type="checkbox"
          />
          HBA
        </label>
        <label>
          <input
            type="checkbox"
          />
          Meat
        </label>
        <label>
          <input
            type="checkbox"
          />
          Meat by #
        </label>
        <label>
          <input
            type="checkbox"
          />
          Produce
        </label>
        <label>
          <input
            type="checkbox"
          />
          Produce by #
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiSelect tests should be callable withough a selectBoxId 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="select-box-id"
    >
      <select
        class="form-control"
        id="selectId"
      >
        <option
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    />
  </div>
</div>
`;

exports[`MultiSelect tests should be callable without a backgroundColor, selectId, or inputId 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="select-box-id"
    >
      <select
        class="form-control"
        id="select-id"
      >
        <option
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    />
  </div>
</div>
`;

exports[`MultiSelect tests should be callable without an id 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="selectBoxId"
    >
      <select
        class="form-control"
        id="selectId"
      >
        <option
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    />
  </div>
</div>
`;

exports[`MultiSelect tests should be clickable 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="select-box-id"
    >
      <select
        class="form-control"
        id="select-id"
      >
        <option
          style="display: none;"
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    >
      <div
        class="multi-select-checkboxes"
        id="multi-select-checkboxes"
        style="top: 0px; left: 0px; background-color: rgb(31, 30, 30);"
      >
        <input
          class="form-control"
          id="multi-select-search"
          placeholder="Search"
          type="text"
          value=""
        />
        <label>
          <input
            type="checkbox"
          />
          Grocery
        </label>
        <label>
          <input
            type="checkbox"
          />
          Nonfood
        </label>
        <label>
          <input
            type="checkbox"
          />
          HBA
        </label>
        <label>
          <input
            type="checkbox"
          />
          Meat
        </label>
        <label>
          <input
            type="checkbox"
          />
          Meat by #
        </label>
        <label>
          <input
            type="checkbox"
          />
          Produce
        </label>
        <label>
          <input
            type="checkbox"
          />
          Produce by #
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiSelect tests should render correctcly 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="selectBoxId"
    >
      <select
        class="form-control"
        id="selectId"
      >
        <option
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    />
  </div>
</div>
`;
